### DISCUSSION TITLE
Run ExApp without Orchestration

### DISCUSSION THREAD

**Discussion by HoffmannP** [nextcloud/app_api#416] [General]
Hy, I'd like to orchestrate my ExApp-Container myself and just register the already running Container as an ExApp to Nextcloud. Is there any way to do this? I thus wouldn't need a Docker Daemon.

**Comment by andrey18106** [comment]
Hello, there is a `manual-install` (https://nextcloud.github.io/app_api/notes_for_developers/DevelopmentEnvironment.html) Deploy daemon type which can be used to register already running ExApp. For examples you can have a look at these PyCharm run configuration to start manually ExApp (https://github.com/cloud-py-api/nc_py_api/tree/main/.run), and then register it via makefile commands (https://github.com/cloud-py-api/nc_py_api/blob/main/Makefile, https://github.com/cloud-py-api/nc_py_api/blob/main/scripts/dev_register.sh).
If there is something missing for your scenario, feel free to open an issue.

**Comment by HoffmannP** [comment]


**Reply by HoffmannP** [reply]
As a documentation for others that may want to run an ExApp without a Docker Orchestration.
You need the following ENV-Variables (everything uppercase):
* NEXTCLOUD_URL (points back to your NC instance)
* APP_VERSION (should be the same as the "version"-value in the xml/json info you use for registering your app)
* APP_ID (as "appid"/"id" above)
* APP_SECRET (required, you must give the same "secret"-value in your xml/json while registering)

**Comment by HoffmannP** [comment]
Also, wenn registering the app manualy via app_api:app:register you need to add the route to your registration information like this:
> app_api:app:register --json-info '{"name":"$NAME","scopes": ["ALL"],"version":"0.1","port":80,"secret":"$SECRET","external-app":{"routes":[{"url":"$ROUTE","verb":"POST","access_level":1}]}}'

(this is the minimum I was able to get to run)
```json
{
    "name": "$NAME",
    "scopes": [
        "ALL"
    ],
    "version": "0.1",
    "port": 80,
    "secret": "$SECRET",
    "external-app": {
        "routes": [
            {
                "url": "$ROUTE",
                "verb": "POST",
                "access_level": "$LEVEL"
            }
        ]
    }
}
```

Where $LEVEL can be one of
* 0: For everybody
* 1: For registered users
* 2: For Administrators

**Comment by stellarpower** [comment]
Just wondering, presumably ExApps currently can only function when inside containers - please correct me if I am wrong, and if not, can I ask the reasons for this design? Personally, I'm already running nextcloud itself inside a pod, so whilst if somoene has it installed on the root then the idea of containers makes sense, as mine already lives in one, I can't help but feel that needing to bring up another container for each extension I want to install seems like a bit of overkill for me as an administrator. I'm looking to make a few small extensions and scripts to perform some automation, etc., patching over some of the holes where forms, flows, and various other addons it seems don't currently give me the ability to do what I'm looking to do. If I have to bring up a separate container once and can bung a lot of stuff in there, then not a major problem, but if I had to run docker-in-podman I would absolutely just give up and not try. I think isolation is a good concept, but without knowing more, as the administrator of my system, (and therefore, I am speaking from my usecase), two levels of container sandboxing does feel like a bit much to install an extension for the one application; it'd be nice to have the option just to install within the one. I do sometimes bridge pods but that's more when I'm connecting nextcloud to something else, rather than installing something extra to work inside nextcloud. Plus, it's slow enough as it is already without any extra overhead of IPC.

**Reply by stellarpower** [reply]
Just bumping this as still relevant. I am writing a script to import events from an ICS file automatically. IMO there seems to be a gap in coverage - the `occ` CLI is designed more for administration and so some tasks can be automated using it, but not all (and this is probably a good thing as that's what it's for). I can perform certain operations using DAV, but then I have to authenticate as if a user rather than as a deamon running on the installation. Flows I never seem to be able to get working and they also seem pretty limited and don't expose all the functionalities of NC, and the apps_scripts addon is helpful but again, not all functionality has been exposed into Lua yet. Something like the ExApps Python API sounds like a great idea but as an administrator I'd like to be able to run a script within my container on a cronjob and with permissions to access everything without orchestrating a whole extra jail to contain it. Exposing a container manager socket is also a security risk, and whilst it makes sense in some setups to use docker as a way to isolate things, I'd add that the majority of docker security problems come from insecure practices like this rather than true vulnerabilities in the implementation and users should always think twice before following any instructions that suggest exposing docker itself to a running container.
